index/mem/index.go

```
@@ -0,0 +1,302 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package mem
+
+import (
+	"crypto/md5"
+	"sync"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index"
+
+	"github.com/RoaringBitmap/roaring"
+	"github.com/uber-go/atomic"
+)
+
+const initialSize = 1024 * 1024
+
+// TODO(prateek): investigate impact of native heap
+type memIndex struct {
+	opts   Options
+	hashFn doc.HashFn
+
+	docIDGen *atomic.Uint32
+
+	// hash(doc.ID)  -> internal docID
+	idsLock sync.RWMutex
+	idsMap  map[doc.Hash]index.DocID
+
+	// internal docID -> document
+	documentsLock sync.RWMutex
+	documents     map[index.DocID]document
+
+	// fieldName -> (Term -> postingID)
+	fieldLock sync.RWMutex
+	fieldMap  map[string]*termsDict
+
+	// postingID -> bitmap of docIDs
+	postingsList postingsList
+}
+
+type document struct {
+	doc.Document
+	docID      index.DocID
+	tombstoned bool
+}
+
+// New returns a new in-memory index.
+func New(opts Options) Index {
+	hashFn := func(i doc.ID) doc.Hash {
+		return md5.Sum([]byte(i))
+	}
+
+	return &memIndex{
+		opts:   opts,
+		hashFn: hashFn,
+
+		docIDGen: atomic.NewUint32(0),
+
+		idsMap:       make(map[doc.Hash]index.DocID, initialSize),
+		documents:    make(map[index.DocID]document, initialSize),
+		fieldMap:     make(map[string]*termsDict, initialSize),
+		postingsList: newPostingsList(initialSize),
+	}
+}
+
+// TODO(prateek): consider copy semantics for input data, esp if we can store it on the native heap
+func (i *memIndex) Insert(d doc.Document) error {
+	// early terminate if id is already present
+	docHash := i.hashFn(d.ID)
+	i.idsLock.RLock()
+	_, ok := i.idsMap[docHash]
+	i.idsLock.RUnlock()
+	if ok {
+		return nil
```

I think we should return an error here. Although the ID is the same, the contents of the document may be different from what is already stored in the index and the caller would have no way of knowing that we didn't store the contents of the document if they did change.

---

Agreed. Will do.

---

done at the Index level (instead of at the Segment level)